EyeOnBuzz/Screens/UpcomingMovies/UpcomingMoviesViewController.swift

```diff
@@ -0,0 +1,187 @@
+        self.genresDataSource = GenresDataSource.init(dataSourceTarget: self)
+        self.upcomingMoviesDataSource = UpcomingMoviesDataSource.init(dataSourceTarget: self)
+        
+        self.tableView.accessibilityLabel = "Upcoming Movies"
+        self.tableView.register(UINib.init(nibName: "UpcomingMovieCell", bundle: nil), forCellReuseIdentifier: self.cellIdentifier)
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        self.setupAppearance()
+        
+        LoadingIndicator.start()
+        
+        self.refreshControl = UIRefreshControl()
+        self.refreshControl?.addTarget(self, action: #selector(updateData), for: UIControlEvents.valueChanged)
+    }
+    
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        
+        if genresDataSource!.list.isEmpty || upcomingMoviesDataSource!.list.isEmpty {
+            self.updateData()
+        }
+    }
+    
+    func subscribe() {
+        NotificationCenter.default.addObserver(self, selector: #selector(updateData), name: .fetchUpcomingMoviesData, object: nil)
+        NotificationCenter.default.addObserver(self, selector: #selector(showInvalidAPIKeyMessage), name: .invalidAPIKey, object: nil)
+        NotificationCenter.default.addObserver(self, selector: #selector(showUnreachableNetworkMessage), name: .unreachableNetwork, object: nil)
+    }
+    
+    // MARK: - Data
+    
+    @objc func updateData() {
+        self.genresDataSource?.fetch()
+    }
+    
+    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        LoadingIndicator.start()
+        
+        let upcomingMovie = self.upcomingMoviesDataSource?.atIndex(indexPath.row) as! UpcomingMovie
+        
+        let upcomingMovieViewController = UpcomingMovieViewController(upcomingMovie)
+        
+        self.navigationController?.pushViewController(upcomingMovieViewController, animated: true)
+        self.navigationItem.backBarButtonItem = BackButton.build()
+    }
+    
+    override func scrollViewDidEndDragging(_ scrollView: UIScrollView, willDecelerate decelerate: Bool) {
+        let currentOffset = scrollView.contentOffset.y
+        let maximumOffset = scrollView.contentSize.height - scrollView.frame.size.height
+        
+        if (maximumOffset - currentOffset) <= 150.0  && !self.upcomingMoviesDataSource!.isLastPage() {
+            LoadingIndicator.start()
+            
+            DispatchQueue.global(qos: .background).async {
+                self.upcomingMoviesDataSource?.fetchNextPage()
+            }
+        }
+    }
+    
+    // MARK: - DataSourceTarget Protocol
+    
+    func dataRefreshed(source: DataSource.RefreshSource, status: DataSource.RefreshStatus) {
+        switch status {
+        case .success:
+            checkSuccessDataSource(source)
```

Comment:
This ☝️  method name is a bit confusing. You just checked the status and it is success and then you call `checkSuccess...` again?

Reply:
Hmmm, its true, I could improve this naming by something like: `checkOrigin`. I decided to extract to a new method to be cleaner. When `success` is called I need to check if was `genre` list response or `upcomingMovies` one. If you want I can give more details why.